opm/common/utility/OpmInputError.hpp

```diff
@@ -0,0 +1,77 @@
+/*
+  Copyright 2020 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <string>
+
+#include <fmt/format.h>
+
+#ifndef OPM_ERROR_HPP
+#define OPM_ERROR_HPP
+
+namespace Opm {
+
+/*
+  The OpmInputError is a custom exception class which can be used to signal
+  errors in input handling. The importance of the OpmInputError exception is
+  *not* the tecnical functionality it provides, but rather the convention
+  surrounding it, when and how it should be used.
+
+  The OpmInputError should be used in situations which are "close to user
+  input", the root cause can either be incorrect user input or a bug/limitation
+  in opm. OpmInputError should only be used in situations where we have a good
+  understanding of the underlying issue, and can provide a good error message.
+
+  The local error handling should be complete when the OpmInputError is
+  instantiated, it should not be caught and rethrown in order to e.g. add
+  additional context or log messages.
+
+  The OpmInputError is not derived from std::exception in to order to facilitate
+  code like this:
+
+     try {
+         // Do some input processing
+     } catch (const std::exception& std_error) {
+         // Catch a standard bug/problem like std::invalid_argument()
+     }
+
+  The standard exceptions are caught and we can e.g. decorate with more context
+  and rethrow, however if OpmInputError has occured that should just pass the
+  catch block and bubble further up.
+*/
+
+
+
+class OpmInputError {
+public:
+    OpmInputError(const std::string& message, const KeywordLocation& loc) :
+        msg_fmt(message),
+        location(loc)
+    {}
+
+    std::string what() const {
+        return fmt::format(this->msg_fmt, this->location.keyword, this->location.filename, this->location.lineno);
```

Comment:
I think this is a bit obscure. In the test example, things like
```
OpmInputError error2("{2}:{1}:{0}", location); 
```
are not possible to understand without reading the code, or at least documenting it. I propose either:
 - leave the formatting to the user, do not take a KeywordLocation argument, or
 - require the KeywordLocation, and adopt a standard scheme for the location information printing, the message string is added verbatim.

Reply:
I'll document it; however I would really like to retain the `KeywordLocation` argument - that clearly signals that the `OpmInputError` should only be used in situations where we have "a keyword to blame" - and a good understanding of the situation.

Reply:
I agree, of the two options I proposed I prefer the second. But I really think we should not pass a string that will be used as a formatting string, that is a bit too obscure I think. Also, with a standard scheme for the info it is easier to grep for in the output.

Reply:
> But I really think we should not pass a string that will be used as a formatting string, that is a bit too obscure I think

Let's give it a shot - I have added documentation to it now.